app/gui2/src/providers/widgetRegistry.ts

```diff
@@ -374,13 +386,26 @@ export class WidgetRegistry {
 
       // Perform a match and update the best widget if the match is better than the previous one.
       const score = widgetModule.widgetDefinition.score(props, this.db)
-      // If we found a perfect match, we can return immediately, as there can be no better match.
-      if (score === Score.Perfect) return widgetModule
+      if (score > Score.Mismatch) {
+        foundLeafMatch ||= widgetModule.widgetDefinition.allowAsLeaf
+      }
       if (score > bestScore) {
         bestScore = score
         best = widgetModule
       }
+
+      // If we found a perfect match, we can return immediately, as there can be no better match.
+      // If we found a leaf match, we can safely return – there will be another leaf later down the tree.
+      // One caveat here is that if the widget marked with `allowAsLeaf = false` has no children. Then we
+      // can pick it here, and it will effectively become leaf.
```

Comment:
I would somehow merge the first two sentences. And I have an impression, that the `caveat` does not explain too much - after all, the `allowAsLeaf` is about guarantees, the widget may do whatever it wants.

Reply:
I don’t think merging is possible without murdering the grammar, but I removed the part about `caveat` and rephrased it.